File: reactive/src/main/java/io/vertx/demos/WSClient.java
package io.vertx.demos;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Vertx;
import io.vertx.core.http.WebSocket;
import io.vertx.core.http.WebSocketFrame;
import io.vertx.core.json.JsonObject;

public class WSClient extends AbstractVerticle {

  public static void main(String[] args) {
    Vertx.vertx().deployVerticle(new WSClient());
  }

  private static final String PING = new JsonObject().put("type", "ping").encode();
  private static final String SUBSCRIBE_TIME = new JsonObject().put("type", "register").put("address", "time").encode();

  @Override
  public void start() {
    // 10k problem, no big deal!
    final WebSocket[] sockets = new WebSocket[20 * 1024];

    for (int i = 0; i < sockets.length; i++) {
      final int id = i;

      vertx.createHttpClient().websocket(8080, "localhost", "/eventbus/websocket", ws -> {
        sockets[id] = ws;
        System.out.println("Connected " + id);

        ws.frameHandler(frame -> System.out.println("Received message: " + frame.binaryData()));

        ws.exceptionHandler(Throwable::printStackTrace);

        ws.endHandler(v -> {
          System.err.println("Connection ended: " + id);
          sockets[id] = null;
        });

        // subscribe to the "time" address
        ws.writeFrame(WebSocketFrame.textFrame(SUBSCRIBE_TIME, true));
      });
    }

    // keep the socket open by sending ping messages every 2:30 min
    vertx.setPeriodic(5 * 60 * 1000 / 2, t -> {
      for (WebSocket socket : sockets) {
        socket.writeFrame(WebSocketFrame.textFrame(PING, true));
      }
    });
  }
}

File: reactive/src/main/resources/wsclient.js
var WebSocketFrame = require('vertx-js/web_socket_frame');

var PING = JSON.stringify({type: 'ping'});
var SUBSCRIBE_TIME = JSON.stringify({type: 'register', address: 'time'});
var len = 10; //10 * 1024;
var sockets = [];

// 10k problem, no big deal!
for (var i = 0; i < len; i++) {
  var id = i;

  vertx.createHttpClient().websocket(8080, "localhost", "/eventbus/websocket", function (ws) {
    sockets[id] = ws;
    console.log("Connected " + id);

    ws.frameHandler(function (frame) {
      console.log("Received message: " + frame.binaryData());
    });

    ws.exceptionHandler(function (e) {
      console.error(e.printStackTrace());
    });

    ws.endHandler(function () {
      console.log("Connection ended: " + id);
      sockets[id] = null;
    });

    // subscribe to the "time" address
    ws.writeFrame(WebSocketFrame.textFrame(SUBSCRIBE_TIME, true));
  });
}

// keep the socket open by sending ping messages every 2:30 min
setInterval(function () {
  for (var i = 0; i < len; i++) {
    if (sockets[i]) {
      sockets[i].writeFrame(WebSocketFrame.textFrame(PING, true));
    }
  }
}, 5 * 60 * 1000 / 2);
